flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java

```diff
@@ -0,0 +1,367 @@
+            sink.write(String.valueOf(i));
+        }
+        assertEquals(20, res.size());
+        assertEquals(Arrays.asList(20, 21, 22), new ArrayList<>(sink.snapshotState().get(0)));
+    }
+
+    @Test
+    public void preparingCommitAtSnapshotTimeEnsuresTheBufferedRecordsArePersistedToDestination()
+            throws IOException, InterruptedException {
+        AsyncSinkWriterImpl sink = new AsyncSinkWriterImpl(sinkInitContext, 10, 1, 100, false);
+        for (int i = 0; i < 23; i++) {
+            sink.write(String.valueOf(i));
+        }
+        sink.prepareCommit(true);
+        assertEquals(23, res.size());
+    }
+
+    @Test
+    public void snapshotsAreTakenOfBufferCorrectlyBeforeAndAfterAutomaticFlush()
+            throws IOException, InterruptedException {
+        AsyncSinkWriterImpl sink = new AsyncSinkWriterImpl(sinkInitContext, 3, 1, 100, false);
+
+        sink.write("25");
+        sink.write("55");
+        assertEquals(Arrays.asList(25, 55), new ArrayList<>(sink.snapshotState().get(0)));
+        assertEquals(0, res.size());
+
+        sink.write("75");
+        assertEquals(Arrays.asList(), new ArrayList<>(sink.snapshotState().get(0)));
+        assertEquals(3, res.size());
+    }
+
+    @Test
+    public void snapshotsAreTakenOfBufferCorrectlyBeforeAndAfterManualFlush()
+            throws IOException, InterruptedException {
+        AsyncSinkWriterImpl sink = new AsyncSinkWriterImpl(sinkInitContext, 3, 1, 100, false);
+        sink.write("25");
+        sink.write("55");
+        sink.write("75");
+        sink.write("95");
+        sink.write("955");
+        assertEquals(Arrays.asList(95, 955), new ArrayList<>(sink.snapshotState().get(0)));
+        sink.prepareCommit(true);
+        assertEquals(Arrays.asList(), new ArrayList<>(sink.snapshotState().get(0)));
+        assertEquals(5, res.size());
+    }
+
+    @Test
+    public void runtimeErrorsInSubmitRequestEntriesEndUpAsIOExceptionsWithNumberOfFailedRequests()
+            throws IOException, InterruptedException {
+        AsyncSinkWriterImpl sink = new AsyncSinkWriterImpl(sinkInitContext, 3, 1, 100, true);
+        sink.write("25");
+        sink.write("55");
+        sink.write("75");
+        sink.write("95");
+        sink.write("35");
+        Exception e = assertThrows(RuntimeException.class, () -> sink.write("135"));
+        assertEquals(
+                "Deliberate runtime exception occurred in SinkWriterImplementation.",
+                e.getMessage());
+        assertEquals(3, res.size());
+    }
+
+    @Test
+    public void retryableErrorsDoNotResultInViolationOfAtLeastOnceSemanticsDueToRequeueOfFailures()
+            throws IOException, InterruptedException {
+        AsyncSinkWriterImpl sink = new AsyncSinkWriterImpl(sinkInitContext, 3, 1, 100, true);
+
+        sink.write("25");
+        assertEquals(Arrays.asList(), res);
+        assertEquals(Arrays.asList(25), new ArrayList<>(sink.snapshotState().get(0)));
```

Comment:
nit: A lot of repetition here, method is long. Consider extracting the 3 lines to a method if it makes it more readable.